virtualizarr/writers/icechunk.py

```diff
@@ -0,0 +1,203 @@
+from typing import TYPE_CHECKING, cast
+
+import numpy as np
+from xarray import Dataset
+from xarray.backends.zarr import encode_zarr_attr_value
+from xarray.core.variable import Variable
+
+from virtualizarr.manifests import ChunkManifest, ManifestArray
+from virtualizarr.zarr import encode_dtype
+
+if TYPE_CHECKING:
+    from icechunk import IcechunkStore  # type: ignore[import-not-found]
+    from zarr import Group  # type: ignore
+
+
+VALID_URI_PREFIXES = {
+    "s3://",
+    # "gs://",  # https://github.com/earth-mover/icechunk/issues/265
+    # "azure://",  # https://github.com/earth-mover/icechunk/issues/266
+    # "r2://",
+    # "cos://",
+    # "minio://",
+    "file:///",
+}
+
+
+def dataset_to_icechunk(ds: Dataset, store: "IcechunkStore") -> None:
+    """
+    Write an xarray dataset whose variables wrap ManifestArrays to an Icechunk store.
+
+    Currently requires all variables to be backed by ManifestArray objects.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+    store: IcechunkStore
+    """
+    try:
+        from icechunk import IcechunkStore  # type: ignore[import-not-found]
+        from zarr import Group  # type: ignore[import-untyped]
+    except ImportError:
+        raise ImportError(
+            "The 'icechunk' and 'zarr' version 3 libraries are required to use this function"
+        )
+
+    if not isinstance(store, IcechunkStore):
+        raise TypeError(f"expected type IcechunkStore, but got type {type(store)}")
+
+    if not store.supports_writes:
+        raise ValueError("supplied store does not support writes")
+
+    # TODO only supports writing to the root group currently
+    # TODO pass zarr_format kwarg?
+    root_group = Group.from_store(store=store)
+
+    # TODO this is Frozen, the API for setting attributes must be something else
+    # root_group.attrs = ds.attrs
+    for k, v in ds.attrs.items():
+        root_group.attrs[k] = encode_zarr_attr_value(v)
+
+    return write_variables_to_icechunk_group(
+        ds.variables,
+        store=store,
+        group=root_group,
+    )
+
+
+def write_variables_to_icechunk_group(
+    variables,
+    store,
+    group,
+):
+    virtual_variables = {
+        name: var
+        for name, var in variables.items()
+        if isinstance(var.data, ManifestArray)
+    }
+
+    loadable_variables = {
+        name: var for name, var in variables.items() if name not in virtual_variables
+    }
+
+    # First write all the non-virtual variables
+    if len(loadable_variables) > 0:
+        # NOTE: We must set the attributes of the group before writing the dataset because the dataset
+        # will overwrite the root group's attributes with the dataset's attributes.
+        # TODO: Is this a sign that something is incorrect?
```

Comment:
That seems like expected behaviour to me? If we did `ds.to_zarr` even if `loadable_variables={}` then that line could double as us setting group attributes.